src/metabase/models/params/chain_filter/dedupe_joins.clj

```diff
@@ -1,102 +1,88 @@
+  The degree of a node is the number of edges incident to it. A loop edge contributes 2 to the degree.
+  Nodes without any edges are not included in the map."
+  [nodes edges]
+  (reduce (fn [degrees edge]
+            (reduce #(update %1 %2 (fnil inc 0))
+                    degrees
+                    (filter nodes (edge-nodes edge))))
+          {}
+          edges))
 
-(defn- list-beforeo
-  "A relation such that `sublist` is all items in `lst` up to (but not including) `item`."
-  [lst sublist item]
-  #_:clj-kondo/ignore
-  (l/matcha [lst sublist]
-    ([[] []])
-    ([[item . _] []])
-    ([[?x . ?list-more] [?x . ?sublist-more]]
-     (list-beforeo ?list-more ?sublist-more item))))
+(defn- make-tree
+  "Return a subset of `edges` forming a tree starting from node `source-id` and
+  containing all nodes in `terminal-ids`.
+  The tree is built greedily, there is no guarantee that we find the optimal one."
+  [source-id edges terminal-ids]
+  (loop [s #{source-id}, edges edges, tree-edges []]
+    (if (set/subset? terminal-ids s)
+      tree-edges
+      (if-let [out-edges (seq (filter (comp s source-node) edges))]
+        (let [edge (apply min-key #(weight terminal-ids %) out-edges)
+              s' (conj s (target-node edge))]
+          (recur s'
+                 (remove (comp s' target-node) edges)
+                 (conj tree-edges edge)))
+        tree-edges))))
 
-(defn- parent-beforeo
-  "A relationship such that the parent join of `join` appears before it in `joins`."
-  [joins join]
-  (l/fresh [joins-before parent]
-    (list-beforeo joins joins-before join)
-    (parent-joino parent join)
-    (l/membero parent joins-before)))
-
-(defn- distinct-rhso
-  "A relationship such that all RHS tables in `joins` are distinct."
-  [joins]
-  (let [rhses (vec (l/lvars (count joins)))]
-    (dorun (map rhso joins rhses))
-    (l/all
-     (l/distincto rhses))))
+(defn- forced-nodes
+  "Return the transitive closure of nodes reachable from `start-nodes` by a single edge only.
+  `node-fn` selects the target node of an edge returned by `node->edges`.  `node-edges` is a function returning the
+  edges going from a node. (Note that source and target nodes can be both lhs and rhs tables depending on `node-fn`
+  and `node-edges`.)"
+  [node-fn start-nodes node->edges]
+  (loop [[node & nodes] start-nodes, result start-nodes]
+    (if (nil? node)
+      result
+      (let [[edge & more] (node->edges node)]
+        (if (or (nil? edge) (seq more))
+          (recur nodes result)
+          (let [next-node (node-fn edge)]
+            (recur (cons next-node nodes) (conj result next-node))))))))
 
 (defn dedupe-joins
   "Remove unnecessary joins from a collection of `in-joins`.
 
   `keep-ids` = the IDs of Tables that we want to keep joins for. Joins that are not needed to keep these Tables may be
-  removed."
+  removed.
+
+  Note that this function implements a simple greedy algorithm, replacing the previous optimal, but exponential
```

Comment:
Do you think the old implementation was slow because it found an optimal solution, or just because `core.logic` is slow?

Reply:
stupid follow-on question: is there anything else that depends on us returning an optimal solution here? I assume the answer is no, but not familiar enough with how these are used to be sure.

Reply:
I think the main reason it was slow is because it set out to solve a difficult problem and it searched an exponential search space. That search could be done faster, it would it still get too slow quickly (an edge or two more later).